### ISSUE TITLE
[Bug] datasets_ingest_train_infer is flaky in the master

### ISSUE THREAD

**Issue by rkooo567** [ray-project/ray#21812]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

We are trying to promote this test to stable, but this seems to be still unstable with runtime error.

Example: 
- https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_fFQFTj58wJZsceMvUMdGtYwq
- https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_AJejf2q5BYQdUXUR28iiJgsT

### Versions / Dependencies

Master

### Reproduction script

N/A

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by scv119** [comment]
```
(BackendExecutor pid=214, ip=10.0.0.163) 2022-01-17 08:21:46,169        ERROR worker.py:86 -- Unhandled error (suppress with RAY_IGNORE_UNHANDLED_ERRORS=1): ray::BaseWorkerMixin._BaseWorkerMixin__execute() (pid=258, ip=10.0.0.209, repr=<ray.train.worker_group.BaseWorkerMixin object at 0x7feefa1b1d10>)
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/train/worker_group.py", line 26, in __execute
(BackendExecutor pid=214, ip=10.0.0.163)     return func(*args, **kwargs)
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/train/backend.py", line 495, in end_training
(BackendExecutor pid=214, ip=10.0.0.163)     output = session.finish()
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/train/session.py", line 103, in finish
(BackendExecutor pid=214, ip=10.0.0.163)     func_output = self.training_thread.join()
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/train/utils.py", line 89, in join
(BackendExecutor pid=214, ip=10.0.0.163)     raise self.exc
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/train/utils.py", line 82, in run
(BackendExecutor pid=214, ip=10.0.0.163)     self.ret = self._target(*self._args, **self._kwargs)
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/train/trainer.py", line 409, in <lambda>
(BackendExecutor pid=214, ip=10.0.0.163)     return lambda: train_func(config)
(BackendExecutor pid=214, ip=10.0.0.163)   File "ray_sgd_training.py", line 343, in train_func
(BackendExecutor pid=214, ip=10.0.0.163)   File "ray_sgd_training.py", line 252, in train_epoch
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
(BackendExecutor pid=214, ip=10.0.0.163)     result = self.forward(*input, **kwargs)
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 511, in forward
(BackendExecutor pid=214, ip=10.0.0.163)     output = self.module(*inputs[0], **kwargs[0])
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
(BackendExecutor pid=214, ip=10.0.0.163)     result = self.forward(*input, **kwargs)
(BackendExecutor pid=214, ip=10.0.0.163)   File "ray_sgd_training.py", line 226, in forward
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
(BackendExecutor pid=214, ip=10.0.0.163)     result = self.forward(*input, **kwargs)
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 91, in forward
(BackendExecutor pid=214, ip=10.0.0.163)     return F.linear(input, self.weight, self.bias)
(BackendExecutor pid=214, ip=10.0.0.163)   File "/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/functional.py", line 1674, in linear
(BackendExecutor pid=214, ip=10.0.0.163)     ret = torch.addmm(bias, input, weight.t())
(BackendExecutor pid=214, ip=10.0.0.163) RuntimeError: mat1 dim 1 must match mat2 dim 0
```

**Comment by scv119** [comment]
Talked with @matthewdeng, we suspect this is caused by the input tensor generated from ray dataset has unexpected shape. for example, it could be the batch size is unexpected, or the num of columns/features is different from what expected. cc @clarkzinzow are you aware of any similar issues?

for now; we will add a check to assert the shape is expected, so if this ever happens we can see the actual shape.

**Comment by clarkzinzow** [comment]
@scv119 @matthewdeng `torch.nn.Linear` is batch-size-agnostic, i.e. it will accept a batch of any size, so I'd expect this to be an issue with a mismatch on the number of columns.

**Comment by matthewdeng** [comment]
Yeah my hunch is that sometimes the internal Arrow column _isn't_ present, but we currently hardcode:
```
    # remove label column and internal Arrow column.
    num_features = num_columns - 2
```

**Comment by clarkzinzow** [comment]
@scv119 This hasn't been deflaked yet, right? We've just added an assertion to get better failure information? If that's the case, then I don't think that this should be closed.

**Comment by scv119** [comment]
right.. i blame the "smart" github that auto closes PRs

**Comment by clarkzinzow** [comment]
So smart! 😂

**Comment by jjyao** [comment]
@scv119 The assertion you added did fail in master: https://buildkite.com/ray-project/periodic-ci/builds/2666#9a83fedc-2454-421b-bd0f-92b9959ac882

**Comment by scv119** [comment]
This has been fixed by #22158
